rfcs/20240919-dnncompat/README.md

```diff
@@ -0,0 +1,936 @@
+        void *workSpace, size_t workSpaceSizeInBytes, const void *beta,
+        const FilterDescriptor_t dwDesc, void *dw);
+
+// The following APIs are used to find the best algorithm given parameters of
+// a convolution operation. When used in dnncompat they will always return 
+// `CONVOLUTION_FWD_ALGO_AUTO` which then gets mapped to 
+// `dnnl::algorithm::convolution_auto` when launching the convolution operation.
+// Thus, dnncompat, relies on oneDNN to select the best algorithm.
+// This behaviour is consistent with vendor-specific libraries, which may return 
+// fewer algorithms than the requested number.
+Status_t findConvolutionForwardAlgorithm(Handle_t,
+        const TensorDescriptor_t, const FilterDescriptor_t,
+        const ConvolutionDescriptor_t, const TensorDescriptor_t, const int,
+        int *returnedAlgoCount, ConvolutionFwdAlgoPerf_t *perfResults);
+
+Status_t getConvolutionForwardAlgorithm_v7(Handle_t,
+        const TensorDescriptor_t, const FilterDescriptor_t,
+        const ConvolutionDescriptor_t, const TensorDescriptor_t, const int,
+        int *returnedAlgoCount, ConvolutionFwdAlgoPerf_t *perfResults);
+
+Status_t findConvolutionBwdDataAlgorithm(Handle_t,
+        const TensorDescriptor_t, const FilterDescriptor_t,
+        const ConvolutionDescriptor_t, const TensorDescriptor_t, const int,
+        int *returnedAlgoCount, ConvolutionBwdDataAlgoPerf_t *perfResults);
+
+Status_t findConvolutionBackwardFilterAlgorithm(Handle_t,
+        const TensorDescriptor_t, const FilterDescriptor_t,
+        const ConvolutionDescriptor_t, const TensorDescriptor_t, const int,
+        int *returnedAlgoCount, ConvolutionBwdFilterAlgoPerf_t *perfResults);
+```
+
+List of enumerations related to the Convolution operation:
+
+```c++
+enum ConvolutionFwdAlgo_t {
+    CONVOLUTION_FWD_ALGO_IMPLICIT_GEMM = 0,
+    CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM,
+    CONVOLUTION_FWD_ALGO_GEMM,
+    CONVOLUTION_FWD_ALGO_DIRECT,
+    CONVOLUTION_FWD_ALGO_FFT,
+    CONVOLUTION_FWD_ALGO_FFT_TILING,
+    CONVOLUTION_FWD_ALGO_WINOGRAD,
+    CONVOLUTION_FWD_ALGO_WINOGRAD_NONFUSED,
+    CONVOLUTION_FWD_ALGO_AUTO,
+    CONVOLUTION_FWD_ALGO_COUNT,
+};
+
+enum ConvolutionBwdDataAlgo_t {
+    CONVOLUTION_BWD_DATA_ALGO_0,
+    CONVOLUTION_BWD_DATA_ALGO_1,
+    CONVOLUTION_BWD_DATA_ALGO_FFT,
+    CONVOLUTION_BWD_DATA_ALGO_FFT_TILING,
+    CONVOLUTION_BWD_DATA_ALGO_WINOGRAD,
+    CONVOLUTION_BWD_DATA_ALGO_WINOGRAD_NONFUSED,
+    CONVOLUTION_BWD_DATA_ALGO_AUTO,
+    CONVOLUTION_BWD_DATA_ALGO_DIRECT,
+    CONVOLUTION_BWD_DATA_ALGO_COUNT,
+};
+
+enum ConvolutionBwdFilterAlgo_t {
+    CONVOLUTION_BWD_FILTER_ALGO_0,
+    CONVOLUTION_BWD_FILTER_ALGO_1,
+    CONVOLUTION_BWD_FILTER_ALGO_FFT,
+    CONVOLUTION_BWD_FILTER_ALGO_3,
+    CONVOLUTION_BWD_FILTER_ALGO_WINOGRAD_NONFUSED,
+    CONVOLUTION_BWD_FILTER_ALGO_FFT_TILING,
+    CONVOLUTION_BWD_FILTER_ALGO_AUTO,
+    CONVOLUTION_BWD_FILTER_ALGO_DIRECT,
+    CONVOLUTION_BWD_FILTER_ALGO_COUNT,
+};
```

Comment:
what should dnncompat API do when facing those algorithms?
In particular, what would SYCLomatic do if the user has some dispatching logic based on those enum values?

For example, what should we do with something like [FindConvolutionForwardAlgorithm](https://docs.nvidia.com/deeplearning/cudnn/latest/api/cudnn-cnn-library.html#cudnnfindconvolutionforwardalgorithmex)?

Reply:
@zhimingwang36, can you please explain how you handle it in SYCLomatic?

Reply:
One possible suggestion would be for the first version of dnncompat if an unsupported algorithm is used to report a warning that it is not supported in oneDNN and fallback to the auto algorithm.